sdk/containers/azcontainerregistry/authentication_policy.go

```diff
@@ -32,93 +29,120 @@ const (
+// Response Body: { refreshToken }
+// Step 4: POST /api/oauth2/token
+// Request Body: { refreshToken, scope, grant-type }
+// Response Body: { accessToken }
+// Step 5: GET /api/v1/acr/repositories
+// Request Header: { Bearer acrTokenAccess }
+// Each registry service shares one refresh token, it will be cached in refreshTokenCache until expire time.
+// Since the scope will be different for different API/repository/artifact, accessTokenCache will only work when continuously calling same API.
 type authenticationPolicy struct {
-	mainResource *temporal.Resource[azcore.AccessToken, acquiringResourceState]
-	cred         azcore.TokenCredential
-	aadScopes    []string
-	acrScope     string
-	acrService   string
-	authClient   *authenticationClient
+	refreshTokenCache *temporal.Resource[azcore.AccessToken, acquiringResourceState]
+	accessTokenCache  string
+	cred              azcore.TokenCredential
+	aadScopes         []string
+	authClient        *authenticationClient
 }
 
 func newAuthenticationPolicy(cred azcore.TokenCredential, scopes []string, authClient *authenticationClient, opts *authenticationPolicyOptions) *authenticationPolicy {
 	return &authenticationPolicy{
-		cred:         cred,
-		aadScopes:    scopes,
-		authClient:   authClient,
-		mainResource: temporal.NewResource(acquire),
+		cred:              cred,
+		aadScopes:         scopes,
+		authClient:        authClient,
+		refreshTokenCache: temporal.NewResource(acquireRefreshToken),
 	}
 }
 
 func (p *authenticationPolicy) Do(req *policy.Request) (*http.Response, error) {
-	// send a copy of the original request without body content
-	challengeReq, err := p.getChallengeRequest(*req)
-	if err != nil {
-		return nil, err
+	var resp *http.Response
+	var err error
+	if req.Raw().Header.Get(headerAuthorization) != "" {
+		// retry request could do the request with existed token directly
+		resp, err = req.Next()
+	} else if p.accessTokenCache != "" {
+		// if there is a previous access token, then we try to use this token to do the request
+		req.Raw().Header.Set(
+			headerAuthorization,
+			fmt.Sprintf("%s%s", bearerHeader, p.accessTokenCache),
+		)
+		resp, err = req.Next()
+	} else {
+		// do challenge process for the initial request
+		var challengeReq *policy.Request
+		challengeReq, err = p.getChallengeRequest(*req)
+		if err != nil {
+			return nil, err
+		}
+		resp, err = challengeReq.Next()
 	}
-	resp, err := challengeReq.Next()
 	if err != nil {
 		return nil, err
 	}
 
-	// do challenge process
+	// if 401 response, then try to get access token
 	if resp.StatusCode == 401 {
-		err := p.findServiceAndScope(resp)
-		if err != nil {
+		var service, scope, accessToken string
+		if service, scope, err = findServiceAndScope(resp); err != nil {
 			return nil, err
 		}
-
-		accessToken, err := p.getAccessToken(req)
-		if err != nil {
+		if accessToken, err = p.getAccessToken(req, service, scope); err != nil {
 			return nil, err
 		}
-
+		p.accessTokenCache = accessToken
```

Comment:
I'm probably missing something here, but shouldn't the access token already be cached in the `temporal.Resource`?

Reply:
That caches the ACR refresh token. The protocol is (@tadelesh please correct my mistakes)
1. acquire AAD access token
1. exchange AAD access token for ACR refresh token
1. exchange ACR refresh token for ACR access token
1. authorize request with ACR access token

`p.accessTokenCache` has the ACR access token. I suppose another `temporal.Resource` could synchronize around that.

Reply:
Thanks for clarifying.

Can the ACR access token be "pushed down" into the existing `temporal.Resource`, or do we need to cache both?

Reply:
I suppose that's possible, however could be more complex than it's worth. First complication that comes to mind is that the access and refresh token could expire separately.

Reply:
Let me add some context. Each container registry shares one refresh token, so one policy created by one client (one client operate one registry) could share and cache this refresh token. With this refresh token, we could get the access token, and DIFFERENT operation (download/upload blob, get/update repository properties, get/update tag, etc.) has DIFFERENT access token. Only when you call two same operation (always happen when you upload blob chunk by chunk, or do some batch operations), the access token can be reused. Meanwhile, refresh token and access token expire separately. So, I use `temporal.Resource` to cache refresh token as it is shared, meanwhile, I use a simple string to cache access token as it is no need to take care of access token's expiration because the access token is large probably invalid for the next request (as I described below). I've added synchronization.